### PR TITLE
[History]: Fix redo after update/publish with transient edits

### LINKED ISSUE #37195: Editor History: Undo doesn't work immediately after publishing / saving
### Description

Undoing immediately after publishing a post in the post editor or saving in the site editor does not work. Undo levels appear to be completely reset.

I can see that this might be deliberate, but what's strange is that adding a few blocks and additional changes after publishing or saving appears to allow undoing / redoing any revision again.

### Step-by-step reproduction instructions

**Block Editor**

1. Navigate to the post editor
2. Add text to a paragraph block
3. In the paragraph block's block settings menu, update line height
4. Update font size
5. Undo until the text and typography updates are gone
6. Redo until you return to the current revision
7. Verify that font-size isn't restored

**Site Editor**

1. Enable full site editing by activating the Gutenberg plugin and an FSE enabled theme (Ex. Quadrat)
2. Navigate to the site editor
3. Add text to a paragraph block
4. In the paragraph block's block settings menu, update line height
5. Update font size
6. Undo until the text and typography updates are gone
7. Redo until you return to the current revision
8. Verify that font-size isn't restored

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ntsekouras**
Fixes: https://github.com/WordPress/gutenberg/issues/37195

Maybe related: https://github.com/WordPress/gutenberg/issues/37760

Investigation started with @youknowriad 's: https://github.com/WordPress/gutenberg/pull/37446#discussion_r777423137.
>I skipped this test because there's an undo/redo bug that is only visible with the empty theme for some reason but not with the tt1-blocks theme.

When we `save` we do not create an undo step in history and was implemented In this [PR](https://github.com/WordPress/gutenberg/pull/17452). With that in place when we have transient edits and update/publish, if we `undo` we cannot `redo` then, as in that case we use the last action(`save`) which is deliberately handled to return early with `fake` no `edits`.

This PR handles this case to follow the normal flow we have for transient edits.

We create transient edits when we update the same block attribute as in the previous action and also when we use `__unstableMarkNextChangeAsNotPersistent` in a block for some initialization purposes in the first render. The latter case was the reason for the failing test with `empty theme`.  In `empty theme`'s index.html file we hadn't set `perPage` and there is [logic there](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query/edit/index.js#L75) to set this value. So when we `reverted the template`, the Query Loop's rendering created transient edits. So after saving we had the above problem.

## Testing instructions
1. I a page type consecutive characters in a `Paragraph` and `update/publish`.
2. Click `redo`
3. Observe that `undo` is available and working as expected

#### Before

https://user-images.githubusercontent.com/16275880/148846443-df4c6c20-c9e8-43cd-a378-059602d593df.mov

 #### After

https://user-images.githubusercontent.com/16275880/148846468-77af7566-c7f1-4c21-a96c-2c9006344ef0.mov

I'd appreciate good testing here for any regression 😄

### CODE REVIEW

File: packages/e2e-tests/specs/site-editor/template-revert.test.js [hunk 1]
```diff
@@ -164,15 +168,14 @@ describe( 'Template Revert', () => {
 		expect( contentBefore ).toBe( contentAfter );
 	} );
 
-	it.skip( 'should show the original content after revert, clicking undo then redo in the header toolbar', async () => {
+	it( 'should show the original content after revert, clicking undo then redo in the header toolbar', async () => {
```
Comment (ntsekouras) [inline packages/e2e-tests/specs/site-editor/template-revert.test.js:171]: Other changes in this file are not relevant to the PR, but refactored a bit (correct function signatures, etc..).

**Review by youknowriad — APPROVED** [review]
While this works, I personally don't understand the logic of that reducer, it's a bit hard to follow.

**Comment by ntsekouras** [comment]
>While this works, I personally don't understand the logic of that reducer, it's a bit hard to follow.

That's so true 😄 . I have other issues about history in my queue and I hope I'll be in a good place soon to try to simplify things there if there is room for it.

**Comment by Mamaduka** [comment]
Sorry, I'm a little bit late. I tested PR, and it works as expected.

> I personally don't understand the logic of that reducer, it's a bit hard to follow.

I also want to echo the same sentiment 😄

**Comment by ntsekouras** [comment]
It feels like we should back port this one, no? --cc @noisysocks

**Comment by youknowriad** [comment]
I added this to the next "minor" as it doesn't seem that critical and I have set another redo bug similarly.
